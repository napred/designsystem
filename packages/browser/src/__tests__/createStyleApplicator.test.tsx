import { css } from 'emotion';
import React from 'react';
import { render } from 'react-testing-library';
import {
  createComponent,
  createSimpleCache,
  createStringStyle,
  createStyle,
  createStyleApplicator,
  DesignSystem,
} from '../';

const Box = createComponent('Box', 'div');

const customStyle = createStyle(['orientation'], (props: any, { theme }) => {
  let orientationCss = css`
    border-bottom-left-radius: 99999px;
    border-bottom-right-radius: 99999px;
  `;

  if (props.orientation === 'top') {
    orientationCss = css`
      border-top-left-radius: 99999px;
      border-top-right-radius: 99999px;
      border-bottom: 1px solid ${theme.color('greyLightest', 'white')};
    `;
  }

  if (props.orientation === 'right') {
    orientationCss = css`
      border-top-right-radius: 99999px;
      border-bottom-right-radius: 99999px;
      border-bottom: 1px solid ${theme.color('greyLightest', 'white')};
    `;
  }

  if (props.orientation === 'left') {
    orientationCss = css`
      border-top-left-radius: 99999px;
      border-bottom-left-radius: 99999px;
      border-bottom: 1px solid ${theme.color('greyLightest', 'white')};
    `;
  }

  return css`
    border: none;
    margin: 0;
    min-width: 0;
    outline: none;
    cursor: pointer;
    transition: background-color ease 0.3s;

    &:hover {
      background-color: ${theme.color('primaryLight')};
    }

    ${orientationCss}
  `;
},['orientation']);

const BoxDistinct = createComponent('BoxDistinct', 'button', { styles: [customStyle] });
BoxDistinct.defaultProps = { 
  bgColor: 'primary',
  color: 'greyLightest',
  display: 'inline-block',
  fontSize: 1,
  height: '2rem',
  lineHeight: 0.1,
  p: 0,
  type: 'button',
  width: '2rem',
 };

const BoxDistinct2 = createComponent('BoxDistinct2', 'div');
const BoxDistinct3 = createComponent('BoxDistinct3', 'span');

describe('createStyleApplicator', () => {
  const styles = [
    createStringStyle('test', 'test'),
    createStyle(
      ['pos'],
      (props: { pos?: string }) => css`
        position: ${props.pos};
      `,
    ),
  ];

  it('works correctly for no props', () => {
    const { asFragment } = render(
      <DesignSystem styleApplicatorFactory={createStyleApplicator} styles={styles}>
        <Box />
      </DesignSystem>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('creates distinct cache for same props but distinct components', () => {
    const cache = createSimpleCache();
    const { asFragment } = render(
      <DesignSystem styleApplicatorFactory={createStyleApplicator} cache={cache} styles={styles}>
        <BoxDistinct />
        <BoxDistinct />
        <BoxDistinct2 />
        <BoxDistinct2 />
        <BoxDistinct3 />
      </DesignSystem>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('works correctly for props and viewport', () => {
    const { asFragment, rerender } = render(
      <DesignSystem styleApplicatorFactory={createStyleApplicator} styles={styles}>
        <Box color="#ff0000" />
        <Box color={['#ff0000', '#000000']} />
      </DesignSystem>,
    );

    const viewport0 = asFragment();

    expect(viewport0).toMatchSnapshot();

    // now change viewport
    rerender(
      <DesignSystem is={1} styleApplicatorFactory={createStyleApplicator} styles={styles}>
        <Box color="#ff0000" />
        <Box color={['#ff0000', '#000000']} />
      </DesignSystem>,
    );

    const viewport1 = asFragment();

    expect(viewport1).toMatchSnapshot();

    expect(viewport1).not.toEqual(viewport0);
  });
});
