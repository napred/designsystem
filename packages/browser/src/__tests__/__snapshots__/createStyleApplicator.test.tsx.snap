// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createStyleApplicator creates distinct cache for same props but distinct components 1`] = `
<DocumentFragment>
  .css-6aoprm {
  border: none;
  margin: 0;
  min-width: 0;
  outline: none;
  cursor: pointer;
  -webkit-transition: background-color ease 0.3s;
  transition: background-color ease 0.3s;
  border-bottom-left-radius: 99999px;
  border-bottom-right-radius: 99999px;
}

.css-6aoprm:hover {
  background-color: #FA4B88;
}

.css-13b38kj {
  background-color: #D03C70;
  color: #FFFFFF;
  display: inline-block;
  font-family: "Palanquin","Helvetica",sans-serif;
  font-size: 20px;
  height: 2rem;
  line-height: 2px;
  padding: 0px;
  width: 2rem;
}

<button
    class="css-6aoprm css-13b38kj"
    type="button"
  />
  .css-6aoprm {
  border: none;
  margin: 0;
  min-width: 0;
  outline: none;
  cursor: pointer;
  -webkit-transition: background-color ease 0.3s;
  transition: background-color ease 0.3s;
  border-bottom-left-radius: 99999px;
  border-bottom-right-radius: 99999px;
}

.css-6aoprm:hover {
  background-color: #FA4B88;
}

.css-13b38kj {
  background-color: #D03C70;
  color: #FFFFFF;
  display: inline-block;
  font-family: "Palanquin","Helvetica",sans-serif;
  font-size: 20px;
  height: 2rem;
  line-height: 2px;
  padding: 0px;
  width: 2rem;
}

<button
    class="css-6aoprm css-13b38kj"
    type="button"
  />
  .css-9ehc2m {
  font-family: "Palanquin","Helvetica",sans-serif;
  font-size: 16px;
  line-height: normal;
}

<div
    class="css-9ehc2m"
  />
  .css-9ehc2m {
  font-family: "Palanquin","Helvetica",sans-serif;
  font-size: 16px;
  line-height: normal;
}

<div
    class="css-9ehc2m"
  />
  .css-9ehc2m {
  font-family: "Palanquin","Helvetica",sans-serif;
  font-size: 16px;
  line-height: normal;
}

<span
    class="css-9ehc2m"
  />
</DocumentFragment>
`;

exports[`createStyleApplicator works correctly for no props 1`] = `
<DocumentFragment>
  .css-9ehc2m {
  font-family: "Palanquin","Helvetica",sans-serif;
  font-size: 16px;
  line-height: normal;
}

<div
    class="css-9ehc2m"
  />
</DocumentFragment>
`;

exports[`createStyleApplicator works correctly for props and viewport 1`] = `
<DocumentFragment>
  .css-j1x65f {
  color: #ff0000;
  font-family: "Palanquin","Helvetica",sans-serif;
  font-size: 16px;
  line-height: normal;
}

<div
    class="css-j1x65f"
  />
  .css-j1x65f {
  color: #ff0000;
  font-family: "Palanquin","Helvetica",sans-serif;
  font-size: 16px;
  line-height: normal;
}

<div
    class="css-j1x65f"
  />
</DocumentFragment>
`;

exports[`createStyleApplicator works correctly for props and viewport 2`] = `
<DocumentFragment>
  .css-j1x65f {
  color: #ff0000;
  font-family: "Palanquin","Helvetica",sans-serif;
  font-size: 16px;
  line-height: normal;
}

<div
    class="css-j1x65f"
  />
  .css-o473nl {
  color: #000000;
  font-family: "Palanquin","Helvetica",sans-serif;
  font-size: 16px;
  line-height: normal;
}

<div
    class="css-o473nl"
  />
</DocumentFragment>
`;
